src/postgres_mcp/server.py

```diff
@@ -16,92 +18,168 @@
 # Constants
 PG_STAT_STATEMENTS = "pg_stat_statements"
 HYPOPG_EXTENSION = "hypopg"
+MAX_RETRIES = 3
+RETRY_DELAY = 1  # seconds
 
 ResponseType = List[types.TextContent | types.ImageContent | types.EmbeddedResource]
 
+logger = logging.getLogger(__name__)
+
+
+def obfuscate_password(url: str) -> str:
+    """Obfuscate password in connection URL."""
+    parsed = urlparse(url)
+    if parsed.password:
+        # Replace password with asterisks
+        netloc = parsed.netloc.replace(parsed.password, "****")
+        return urlunparse(parsed._replace(netloc=netloc))
+    return url
+
 
 class DatabaseConnection:
-    """Database connection manager."""
+    """Database connection manager with retry logic."""
 
     def __init__(self, connection_url: Optional[str] = None):
         self.connection_url = connection_url
         self.conn = None
+        self._is_valid = False
+        self._last_error = None
 
     async def connect(self, connection_url: Optional[str] = None) -> Any:
-        """Connect to the database."""
+        """Connect to the database with retry logic."""
         url = connection_url or self.connection_url
+        self.connection_url = url
         if not url:
-            raise ValueError("Database connection URL not provided")
+            self._is_valid = False
+            self._last_error = "Database connection URL not provided"
+            raise ValueError(self._last_error)
+
+        for attempt in range(MAX_RETRIES):
+            try:
+                self.conn = await psycopg.AsyncConnection.connect(url)
+                # Validate connection
+                async with self.conn.cursor() as cursor:
+                    await cursor.execute("SELECT 1")
+                self._is_valid = True
+                self._last_error = None
+                return self.conn
+            except Exception as e:
+                self._is_valid = False
+                self._last_error = str(e)
+                if attempt < MAX_RETRIES - 1:
+                    logger.warning(
+                        f"Connection attempt {attempt + 1} failed: {obfuscate_password(str(e))}. Retrying in {RETRY_DELAY} seconds..."
+                    )
+                    await asyncio.sleep(RETRY_DELAY)
+                else:
+                    logger.error(f"All connection attempts failed: {obfuscate_password(str(e))}")
+                    raise
+
+    async def _attempt_reconnect(self) -> Any:
+        """Attempt to reconnect to the database."""
+        if not self.connection_url:
+            raise ValueError(
+                f"Database connection is broken and no connection URL is available. Last error: {self._last_error}"
+            )
+        await self.connect()
+        return self.conn
+
+    async def get_connection(self) -> Any:
```

Comment:
Why do we have a `connect` and `get_connection`. Does it make sense?

Reply:
Thanks. Good suggestion. Refactored the code and now it's not needed to have both.